### PR TITLE
Tracked questions

### PR DESCRIPTION
**Pull request by knowuh**
Changes to support adding tracked-questions to activities and sequence.

This branch includes changes to the author system that allow for tracked questions, and reporting API endpoints to search for, and report on answers to tracked questions.

It isn't ready for review yet.  I still need simplify the tracked question report responses, and update some tests.

### CODE REVIEW

**Comment by knowuh** [comment]
Sorry this is large.  Most changes are new react views / controllers / models for creating the tracked question.  The API for querying tracked questions ended up being pretty small / simple.  Please review when you get a chance @scytacki

**Comment by dougmartin** [comment]
I can take a look at this too this morning.

File: app/assets/javascripts/components/question_tracker/image_question.js.coffee [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+      updateDrawingPrompt: (v) ->
+        @setState
+          drawing_prompt:v
+          , => @throttle(200, @save)
+
+      updateBgUrl: (evt) ->
+        @setState
+          bg_url:evt.target.value
+        , @save
+
```
Comment (dougmartin) [inline app/assets/javascripts/components/question_tracker/image_question.js.coffee]: Just curious as to why you throttle the two calls above and not this one or the one below?

File: app/assets/javascripts/components/question_tracker/master_question.js.coffee [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+
+      render: ->
+        question = @props.question
+        details = question.question
+        edit = @props.edit
+        update = @update
+
+        (div {className: 'master-question'},
+          (QuestionAdder {question: question, update: update, edit: edit})
+          if @props.question
```
Comment (dougmartin) [inline app/assets/javascripts/components/question_tracker/master_question.js.coffee]: You have a mix of question and @props.question and edit and @props.edit.  No bigee.

File: app/assets/javascripts/components/question_tracker/question_adder.js.coffee [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+          action:
+            type: "replaceMaster"
+            value: selectValue
+            lastValue: @props.question
+        @props.update newValue
+
+      render: ->
+        question = @props.question
+        update = @update
+        can_update = @props.edit
```
Comment (dougmartin) [inline app/assets/javascripts/components/question_tracker/question_adder.js.coffee]: Why not make this?

     disabled = not @props.edit

File: app/assets/javascripts/components/question_tracker/text_input.js.coffee [hunk 4]
```diff
@@ -0,0 +1,17 @@
+{div, input, label} = React.DOM
+
+modulejs.define 'components/question_tracker/text_input',
+  [],
+  () ->
+
+    React.createClass
+      update: (evt) ->
+        change = {"#{@props.name}": evt.target.value}
+        console.log change
```
Comment (dougmartin) [inline app/assets/javascripts/components/question_tracker/text_input.js.coffee]: Looks like you left in a console.log during debugging.

File: app/controllers/question_trackers_controller.rb [hunk 5]
```diff
@@ -0,0 +1,79 @@
+# encoding: UTF-8
+
+class QuestionTrackersController < ApplicationController
+  before_filter :set_question_tracker, except: [:new, :index, :create]
+  before_filter :record_qt_origin
+
+  def index
+    authorize! :manage, QuestionTracker
+    @question_trackers = QuestionTracker.all # TODO Limit by user.
```
Comment (dougmartin) [inline app/controllers/question_trackers_controller.rb]: Does the TODO mentioned above need to be done for this commit?

File: app/models/question_tracker.rb [hunk 6]
```diff
@@ -0,0 +1,57 @@
+# encoding: UTF-8
+
+class QuestionTracker < ActiveRecord::Base
+  attr_accessible :master_question, :questions, :name, :description, :user
+
+  belongs_to :master_question, polymorphic: true
+  has_many :tracked_questions
+
+  def self.list_tracked_questions(user)
+    # TODO: search by user, other criteria?
```
Comment (dougmartin) [inline app/models/question_tracker.rb]: Does this TODO need to be todone?

**Comment by dougmartin** [comment]
The CoffeeScript part looks good to me.  I made a few nits in the React files - I think the only one that might need a change is removing the console.log.  I looked through the Rails code but @scytacki would be the better reviewer for that.

**Comment by knowuh** [comment]
Thanks for the review @dougmartin .  I will clean that stuff up.

File: app/assets/javascripts/components/question_tracker/image_question.js.coffee (hunk 1 above)
Comment (knowuh) [inline app/assets/javascripts/components/question_tracker/image_question.js.coffee]: The text input listen to on-blur.  The prompts listen to keyup events, which fire all the time.

File: app/controllers/question_trackers_controller.rb (hunk 5 above)
Comment (knowuh) [inline app/controllers/question_trackers_controller.rb]: I think we have to punt on these TODOs for now.  I think they should stay in. Maybe we could calculate technical debt by counting all the TODO tags in a source tree.

**Comment by scytacki** [comment]
@knowuh can you take a look this morning? I'm not sure if github gives you a nice view of just my changes. If not here is the diff:
https://github.com/concord-consortium/lara/compare/a4c597c...fb0241e
There are some notes in the commit messages about the rational.

**Comment by knowuh** [comment]
This looks good to me @scytacki -- I will let you merge it.

**Pull request merged**